src/rust/engine/src/externs/interface.rs

```diff
@@ -566,6 +576,83 @@ py_class!(class PyNailgunServer |py| {
     }
 });
 
+py_class!(class PyNailgunClient |py| {
+  data executor: PyExecutor;
+  data port: u16;
+
+  def execute(&self, py_signal_fn: PyObject, command: String, args: Vec<String>, env: PyDict) -> CPyResult<PyInt> {
```

Comment:
Consider moving the body of this method into a private function just above, e.g.
```rust
fn _execute_nailgun_client(client: &PyNailgunClient, py_signal_fn: Value, command: String, args: Vec<String>, env: Vec<String, String>) -> i64 { ... }
```

The benefit of that is that it would decouple the "massage cpython objects into rust" part from the nailgun execution. Since this is within a macro definition, it seems like there is benefit to reducing the amount of logic defined inline. Totally subjective.

Reply:
This is probably fine, see Stu's comment on argument transforms from cpython here: https://github.com/pantsbuild/pants/pull/10865#discussion_r498967540.